### PR TITLE
Speed up

### PR DESCRIPTION
**Pull request by daikitag**
Previous code had 2 tree traversals, so reduce it to be 1. Also modify sections where the code was inefficient.

### CODE REVIEW

File: tstrait/genetic_value.py [hunk 1]
```diff
@@ -6,6 +9,21 @@
 from .base import _check_instance, _check_dataframe, _check_non_decreasing  # noreorder
 
 
+@dataclass
+class GeneticResult:
+    """Data class that contains effect size dataframe and phenotype dataframe.
+
+    :param effect_size: Effect size dataframe that includes site_id, causal_state,
+        effect_size, and trait_id
+    :param phenotype: Phenotype dataframe that includes trait_id, individual_id,
```
Comment (jeromekelleher) [inline tstrait/genetic_value.py:18]: These param documentation is incorrect.

Also probably better to document the members individually like

```python
effect_size: pd.DataFrame
"""
Effect size dataframe ...
"""
```

File: tstrait/genetic_value.py [hunk 2]
```diff
@@ -69,7 +121,7 @@ def _individual_genotype(self, tree, site, causal_state, num_nodes):
             if state == causal_state:
                 stack.append(node)
 
-        if len(stack) == 0:
+        if len(stack) == 0:  # pragma: no cover
```
Comment (jeromekelleher) [inline tstrait/genetic_value.py:124]: Better to cover this corner case with a test?

**Review by jeromekelleher — COMMENTED** [review]
Looks good. I'm not quite following the refactoring, perhaps you can talk me through tomorrow.

File: tests/test_genetic_value.py [hunk 3]
```diff
@@ -209,79 +264,178 @@ def test_diff_ind_tree(self):
...
 
-        pd.testing.assert_frame_equal(genetic_df, result_df, check_dtype=False)
+        pd.testing.assert_frame_equal(
+            effect_size_df, genetic_result.effect_size, check_dtype=False
+        )
+        pd.testing.assert_frame_equal(
+            genetic_df, genetic_result.genetic, check_dtype=False
+        )
+
+    def test_allele_freq_one(self, sample_trait_model):
```
Comment (daikitag) [inline tests/test_genetic_value.py:405]: With https://github.com/tskit-dev/tstrait/pull/65#discussion_r1297128889, I examined the case in the site where there is only 1 state (reverse mutation completely cancelled out the mutation), and I got the desired output.

File: tstrait/genetic_value.py (hunk 1 above)
Comment (daikitag) [reply tstrait/genetic_value.py:18]: I will modify the docstrings by tomorrow.

**Review by jeromekelleher — APPROVED** [review]
